api/src/pcapi/routes/pro/stocks.py

```diff
@@ -26,6 +28,23 @@
 logger = logging.getLogger(__name__)
 
 
+def _stock_exists(offer_id: int, stock_payload: serialization.StockCreationBodyModel) -> bool:
+    if not stock_payload.price:
+        price_to_check = decimal.Decimal(0)
+    else:
+        price_to_check = stock_payload.price
+    existing_stock = offers_models.Stock.query.filter(
+        offers_models.Stock.offerId == offer_id,
+        offers_models.Stock.isSoftDeleted == False,
+        offers_models.Stock.beginningDatetime == stock_payload.beginning_datetime,
+        offers_models.Stock.bookingLimitDatetime == stock_payload.booking_limit_datetime,
+        offers_models.Stock.price == price_to_check,
+        offers_models.Stock.priceCategoryId == stock_payload.price_category_id,
+        offers_models.Stock.quantity == stock_payload.quantity,
+    ).all()
+    return bool(existing_stock)
```

Comment:
Ça n’est pas plutôt le rôle d’une contrainte en DB ?

Reply:
C'est vrai que là on fait une requête par stock à créer. Ce qui peut faire jusqu'à 2500 requête pour une requête à l'api. 

Si on veut vraiment vérifier qu'un stock n'est pas déjà existant en base je pense qu'on peut le faire d'un coup avant. Au début de la route faire une seule requête pour savoir si un stock existe déjà. 
On pourrait trouver avant coup quels stocks à créer existent déjà et les ignorer

Reply:
La contrainte d'unicité n'est pas possible parce que certains champs sont nullables, comme `beginningDatetime` et `bookingLimitDatetime`

Pour rappel, en SQL `null` est différent de `null` sauf dans un `select distinct`

Reply:
Voici une piste pour éviter les N+1 requêtes : https://chat.openai.com/share/7f809ea3-cfcf-4ac2-80b9-1197feff7f1b

Reply:
@cepehang 

Alors en effet je pensais qu’on pouvait faire quelque chose comme

`CREATE UNIQUE INDEX ... NULLS NOT DISTINCT` mais c’est pas dispo avant postgres15 :x

Reply:
Est-ce qu'on ne pourrait pas régler ce problème avec une contrainte pour les cas où il n'y a pas de dates et une pour les autres lignes ?

Reply:
De base je me disais que c’était certainement faisable de faire:

`CREATE UNIQUE INDEX unique_stock_per_offer ON stock ("offerId", "beginningDatetime", "bookingLimitDatetime", "price", "priceCategoryId", "quantity") WHERE stock."isSoftDeleted" = False`

Pour pouvoir avoir un index partiel et gérer la contrainte d’unicité sur les stocks qui ne sont pas soft delete. Mais en l’état je ne vois pas comment gérer les stocks avec une valeur nullable